python-lib/dku_webapp/utils.py

```diff
@@ -0,0 +1,182 @@
+import dataiku
+import pandas as pd
+import numpy as np
+import logging
+from flask import request
+import simplejson
+import traceback
+from dku_model_accessor import get_model_handler, ModelAccessor
+from dku_model_fairness_report import ModelFairnessMetricReport, ModelFairnessMetric
+from dku_model_fairness_report.constants import DkuFairnessConstants
+from dku_webapp.constants import DkuWebappConstants
+
+
+
+def get_histogram_data(y_true, y_pred, y_pred_proba, advantageous_outcome, sensitive_feature_values):
+    df = pd.DataFrame()
+    # the following strings are used only here, too lazy to turn them into constant variables
+    df['bin_index'] = np.digitize(y_pred_proba, np.arange(0, 1, step=0.1))  # 10 bins
+    df['prediction_result_type'] = get_prediction_result_type(y_true, y_pred, advantageous_outcome)
+    df['sensitive_feature'] = sensitive_feature_values
+
+    cast_to_int = False
+    try: # check whether or not the column can be casted to int
+        if np.array_equal(df['sensitive_feature'], df['sensitive_feature'].astype(int)):
+            cast_to_int = True
+    except:
+        pass
+
+    histogram_dict = {}
+    for v in df['sensitive_feature'].unique():
+
+        df2 = df[df['sensitive_feature'] == v]
+        dfx = np.round(100 * df2.groupby(['prediction_result_type', 'bin_index']).size() / len(df2), 3)
+        series_final = dfx.unstack().fillna(0).stack()
+
+        computed_df = pd.DataFrame(series_final, columns=['bin_value_new'])
```

Comment:
I would try to extract several steps from this code explain through comments what you are doing

Reply:
Done in dbccbda